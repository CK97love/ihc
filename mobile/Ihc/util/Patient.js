// Realm returns arrays as objects, so convert them before sending to server
export function convertObjectsToArrays(patient) {
  const copy = Object.assign({}, patient);
  copy.medications = Array.from(patient.medications);
  copy.soaps = Array.from(patient.soaps);
  copy.triages = Array.from(patient.triages);
  copy.statuses = Array.from(patient.statuses);
  return copy;
}
