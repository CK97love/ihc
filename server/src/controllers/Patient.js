import PatientModel from '../models/Patient';
import SoapModel from '../models/Soap';
import StatusModel from '../models/Status';
import TriageModel from '../models/Triage';
import DrugUpdateModel from '../models/DrugUpdate';

const PatientController = {
  GetPatient: function(req, res){
    PatientModel.findOne({key: req.params.key}, function(err, patient) {
      if(!patient) {
        err = new Error("Patient with key " + req.params.key + " doesn't exist");
      }
      if(err) {
        res.json({status: false, error: err.message});
        return;
      }
      res.json({status: true, patient: patient});
    });
  },
  GetPatients: function(req, res){
  },
  CreatePatient: function(req, res){
  },
  UpdatePatient: function(req, res){
  },
  GetUpdates: function(req, res){
  },
  GetSoap: function(req, res){
  },
  GetStatus: function(req, res){
  },
  GetTriage: function(req, res){
  },
  GetDrugUpdates: function(req, res){
  },
  UpdateSoap: function(req, res){
  },
  UpdateStatus: function(req, res){
  },
  UpdateTriage: function(req, res){
  },
  UpdateDrugUpdates: function(req, res){
  }
};

module.exports = PatientController;
