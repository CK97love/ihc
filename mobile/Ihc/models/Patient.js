import {stringDate} from '../util/Date';
export default class Patient {
  // Insert any class methods here
  get drugUpdates() {
    const drugToUpdates = {};
    for (var update of this.medications) {
      if (update.name in drugToUpdates) {
        drugToUpdates[update.name].push(update);
      } else{
        drugToUpdates[update.name] = [update];
      }
    }
    return drugToUpdates;
  }

  static fullName(patient) {
    return `${patient.firstName} ${patient.fatherName} ${patient.motherName}`;
  }

  // To be used as primary key
  static makeKey(patient) {
    const str = `${patient.firstName}&${patient.fatherName}&${patient.motherName}` +
      `&${patient.birthday}`;
    return str;
  }

  static extractFromForm(form) {
    const patient = Object.assign({}, form);
    patient.birthday = stringDate(form.birthday);
    patient.key = Patient.makeKey(patient);
    if(form.newPatient) {
      // 1 is male, 2 is female
      patient.gender = form.gender === 'Male' ? 1 : 2;
      patient.lastUpdated = new Date().getTime();
    }
    return patient;
  }
}

Patient.schema = {
  name: 'Patient',
  primaryKey: 'key',
  properties: {
    key: 'string',
    firstName: 'string',
    fatherName: 'string', // last name
    motherName: 'string', // last name
    birthday: 'string',
    gender: 'int', // 1 = boy, 2 = girl, 0 = undefined
    phone: 'string?',
    motherHeight: 'double?',
    fatherHeight: 'double?',
    medications: 'DrugUpdate[]',
    soaps: 'Soap[]',
    triages: 'Triage[]',
    statuses: 'Status[]',
    growthchart: 'GrowthChartUpdate[]',
    lastUpdated: 'int' // timestamp
  }
};
