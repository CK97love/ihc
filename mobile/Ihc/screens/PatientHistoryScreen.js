import React, { Component } from 'react';
import {
  Platform,
  StyleSheet,
  TouchableOpacity,
  Text,
  View,
} from 'react-native';
import { Col, Row, Grid } from "react-native-easy-grid";
import data from '../services/DataService';
import Patient from '../models/Patient';

export default class PatientHistoryScreen extends Component<{}> {
  /*
   * Expects:
   *  {
   *    name: string, patient's name (for convenience)
   *    patientKey: string
   *  }
   */
  constructor(props) {
    super(props);
    this.state = {
      patient: {},
      //rows: [],
      soaps: [],
      triages: [],
      loading: false,
      error: null
    };
  }

  loadPreviousVisits = () => {
    this.setState({ loading: true });
    data.getPatient(this.props.patientKey)
      .then( data => {
        this.setState({ patient: data, soaps: data.soaps, triages: data.triages, loading: false });
        //for (i = 0; i < data.soaps.length; i ++) {
          //this.state.rows[i] = data.soaps[i].date;
        //}
      })
      .catch( err => {
        this.setState({ error: err, loading: false });
      });
  }
/*
  loadPreviousVisits = () => {
    this.setState({ loading: true });
    data.getSoap(patientKey, ) 
      .then( data => {
        this.setState({ rows: data, loading: false });
      })
      .catch( err => {
        this.setState({ error: err, loading: false });
      });
  }
*/
  loadAll() {
    this.loadPreviousVisits();
  }

  goToSoap = () => {
    this.props.navigator.push({
      screen: 'Ihc.PatientHistoryScreen',
      title: 'Back to patient',
      passProps: { name: this.props.name, patientKey: this.props.patientKey }
    });
  }

  goToTriage = () => { 
    this.props.navigator.push({
      screen: 'Ihc.PatientHistoryScreen',
      title: 'Back to patient',
      passProps: { name: this.props.name, patientKey: this.props.patientKey }
    });
  }
/*
  renderRow = (data, keyFn) => {
    const cols = data.map( (e,i) => (
      <View style={styles.col} key={keyFn(i)}>
        {( () => {
          switch(i) {
            case 1: //date
              return <Text>{e.
            case 2: //soap
            case 3: //traige
          }
        })() }
      </View>
    ) );
  }
*/
  render() {
    return (
      <View style={styles.container}>
        <Text style={styles.title}>
          Previous Visits
        </Text>

        <View style={styles.gridContainer}>
          <Grid>
            <Col style={styles.col}>
              <Text>
                {this.state.patient.name}
              </Text>
            </Col>

            <Col style={styles.col}>
              <TouchableOpacity style={styles.buttonContainer}
                  onPress={this.goToSoap}>
                <Text style={styles.button}>SOAP</Text>
              </TouchableOpacity>
            </Col>

            <Col style={styles.col}>
              <TouchableOpacity style={styles.buttonContainer}
                  onPress={this.goToTriage}>
                <Text style={styles.button}>Triage</Text>
              </TouchableOpacity>
            </Col>
          </Grid>
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  gridContainer: {
    flex: 1,
    maxWidth: '80%',
    alignItems: 'center',
  },
  col: {
    alignItems: 'center',
  },
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  title: {
    fontSize: 20,
    textAlign: 'center',
    margin: 5,
  },
  buttonContainer: {
    width: 150,
    margin: 10,
    padding: 8,
    elevation: 4,
    borderRadius: 2,
    backgroundColor: '#2196F3',
  },
  button: {
    fontWeight: '500',
    color: '#fefefe',
    textAlign: 'center',
  }
});
