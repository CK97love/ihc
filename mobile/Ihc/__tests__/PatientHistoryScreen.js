import 'react-native';
import React from 'react';
import PatientHistoryScreen from '../screens/PatientHistoryScreen';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import sinon from 'sinon';
import data from '../services/DataService';
import Patient from '../models/Patient';

it('renders correctly', () => {
  sinon.stub(data, 'getPatient').returns(Promise.resolve(Patient.getInstance()));

  const json = renderer.create(
    <PatientHistoryScreen todayDate = '20180101' />
  ).toJSON();
  expect(json).toMatchSnapshot();
});
