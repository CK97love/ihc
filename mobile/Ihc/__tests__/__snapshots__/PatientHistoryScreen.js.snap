// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 20,
        "margin": 5,
        "textAlign": "center",
      }
    }
  >
    Previous Visits
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Loading...
  </Text>
</View>
`;
